Support additional characters in attribute selectors

jordwest opened the pull request:
This change allows `.find('[data-attr1]')` and `.find('[data-attr_foo]')` to match as expected.

According to the HTML5 data-* spec, digits are allowed in data-* attributes.

Additionally, this prevents matching of invalid attributes, such as `:foo`, `-foo`, or `123-foo`,

ljharb commented on test/ReactWrapper-spec.js:
```diff
@@ -369,12 +369,13 @@ describeWithDOM('mount', () => {
 
       expect(wrapper.find('[data-foo="bar"]')).to.have.length(1);
       expect(wrapper.find('[data-foo]')).to.have.length(1);
+      expect(wrapper.find('[data-foo123]')).to.have.length(1);
```
can you add another test for attributes that start with a digit? I'm pretty sure `123-abc` isn't a valid attribute (but either way we should have a test for it)

jordwest replied:
Sure thing, is that `[data-123-abc]`? Or just `[123-abc]`?

ljharb replied:
`data-123-abc` should be valid, i think `123-abc` should be invalid.

jordwest commented:
I've added tests to ensure that `data-123-foo` is valid, and that the following are invalid:
 - `123-foo`
 - `-foo`
 - `:foo`

The last two were failing in master, but a change to the regexp now ensures the first character of an attribute is always an alphabet character.

blainekasten commented on test/ReactWrapper-spec.js:
```diff
@@ -366,15 +366,34 @@ describeWithDOM('mount', () => {
       expect(wrapper.find('[data-foo="bar"]')).to.have.length(1);
       expect(wrapper.find('[data-foo]')).to.have.length(1);
+      expect(wrapper.find('[data-foo-123]')).to.have.length(1);
+      expect(wrapper.find('[data-123-foo]')).to.have.length(1);
```
can you add tests to ensure the value matching still works out correct?

Things like:

```
find('[data-foo-123=3'])
find('[data-foo-123="3"'])
```

Aweary commented on test/ReactWrapper-spec.js:
```diff
@@ -366,15 +366,34 @@ describeWithDOM('mount', () => {
+          '-foo': 'bar',
+          ':foo': 'bar',
+        }))
+      );
```
Can you just use JSX here to be consistent with the other tests? Thanks!

Aweary commented:
@jordwest can you add support for underscores as well? React will render this `<div/>` with the corrent attribute just fine:

```js
<div data-foo_bar='true' />
```
but this query currently fails: 

```js
expect(wrapper.find('[data-foo_bar="true"]')).to.have.length(1);
```

jordwest commented:
@blainekasten:
> can you add tests to ensure the value matching still works out correct?

Added in 31d0ec0

@Aweary:
> can you add support for underscores as well?

Added in 4c60f7b

Aweary commented:
Thanks @jordwest! LGTM
